### PR TITLE
Don't report any errors in `lower_intrinsics`.

### PR DESCRIPTION
**Pull request by oli-obk**
Intrinsics should have been type checked earlier.

This is part of moving all mir-opt diagnostics early enough so that they are reliably emitted even in check builds: https://github.com/rust-lang/rust/issues/49292#issuecomment-1692212095

### CODE REVIEW

File: compiler/rustc_borrowck/src/type_check/mod.rs [hunk 1]
```diff
@@ -1546,33 +1546,44 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
...
-            Some(func.ty(body, self.infcx.tcx))
-        } else {
-            None
-        };
+        let func_ty = func.ty(body, self.infcx.tcx);
+        if let ty::FnDef(def_id, _) = *func_ty.kind() {
+            if self.tcx().is_intrinsic(def_id) {
+                match self.tcx().item_name(def_id) {
+                    sym::simd_shuffle => {
+                        if !matches!(args[2], Operand::Constant(_)) {
```
Comment (bjorn3) [inline compiler/rustc_borrowck/src/type_check/mod.rs:1569]: I believe there are a couple of cases where stdarch has a shuffle where the last arg is not an `Operand::Constant`, but is still constant after LLVM level constprop. I do want this to be fixed though. Handling this in cg_clif requires a bunch of non-trivial code: https://github.com/bjorn3/rustc_codegen_cranelift/blob/0559de65672243a97d6cd6e6ee6a8e8c291ef4ce/src/constant.rs#L476C1-L575 (called from https://github.com/bjorn3/rustc_codegen_cranelift/blob/0559de65672243a97d6cd6e6ee6a8e8c291ef4ce/src/intrinsics/simd.rs#L171-L172) AFAIK stdarch's test suite doesn't get run in rust's CI. Only in the CI of rust-lang/stdarch.

Comment (oli-obk) [reply compiler/rustc_borrowck/src/type_check/mod.rs:1569]: I just moved the check. We were already checking this. I think the const propagated ones haven't been working for a long time

Comment (bjorn3) [reply compiler/rustc_borrowck/src/type_check/mod.rs:1569]: Looks like it is indeed unreachable for `simd_shuffle` now, but for `simd_insert` and `simd_extract` it is still possible and necessary for stdarch.

Comment (oli-obk) [reply compiler/rustc_borrowck/src/type_check/mod.rs:1569]: every `simd_insert` usage outside of ui/codegen tests seems to be using constants

for `simd_extract`, everything but some extract helpers for https://github.com/rust-lang/rust/issues/60637 are using constants

Comment (oli-obk) [reply compiler/rustc_borrowck/src/type_check/mod.rs:1569]: and even those helpers are only used in tests

**Pull request merged**